### ISSUE TITLE
Blob storage capability in state stores: API design

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/dapr#4934]
This is a design proposal for adding support for a set of "blob storage" APIs to state stores. These APIs are an optional capability that state store components can implement, but they will not be required to (just like querying is an optional capability, for example).

See #4804 for an initial discussion on the topic. After an initial proposal that was calling for a brand-new building block, we agreed that this could be implemented as an (optional) capability in state stores, to avoid duplication and reduce maintenance burden. However, it will be exposed to clients using new API endpoints.

## Goal

We define **blobs** as sequences of unstructured data, that should be assumed to be possibly large (many MBs or even GBs in size) and binary. Examples include images and videos, Office documents, etc.

We want to extend the APIs offered by the state store building block to include a new optional "blob storage" capability that includes two methods:

- Retrieving a blob
- Storing a blob
- Delete a blob

These methods are **stream-first**, as they will be working with input and output data as binary streams, and will not perform any transformation or encoding on the data. Thanks to using streams, there are no limits on the size of the input/output data, bypassing `MaxBodySize`.

> This proposal is dependant on the underlying streaming capabilities included in #4903 (which, among other things, updates the HTTP API server to be able to work with streamed data).

## New interface for state stores

The first implementation step requires adding a new optional interface to state stores. State stores must comply with the [`Store` interface](https://github.com/dapr/components-contrib/blob/3b96384de90d6ba7547b88ce920575fe7c443f23/state/store.go#L23) and already have the ability to implement two optional interfaces: [`BulkStore`](https://github.com/dapr/components-contrib/blob/3b96384de90d6ba7547b88ce920575fe7c443f23/state/store.go#L42) and [`Querier`](https://github.com/dapr/components-contrib/blob/3b96384de90d6ba7547b88ce920575fe7c443f23/state/store.go#L98).

We are going to add a new **`BlobStore`** interface that will be optional:

```go
type BlobStore interface {
    // GetBlob retrieves a blob with name key and writes it to the out stream.
    // If the blob has not been found, err will be ErrNotFound.
    // Return value tagOut contains an etag or a similar concurrency-control tag.
    GetBlob(ctx context.Context, key string, out io.Writer) (md BlobMetadata, tagOut any, err error)
    // SetBlob stores a blob with name key, reading data from the in stream.
    // Parameter tag is optional and can be used to pass an etag or any similar concurrency-control tag.
    // In case of conflict, the method returns ErrConflict.
    SetBlob(ctx context.Context, key string, in io.Reader, tag any, md BlobMetadata) (tagOut any, err error)
    // DeleteBlob deletes a blob with name key.
    // Parameter tag is optional and can be used to pass an etag or any similar concurrency-control tag.
    // In case of conflict, the method returns ErrConflict.
    DeleteBlob(ctx context.Context, key string, tag any) (err error)
}

// Metadata associated with a blob.
type BlobMetadata map[string]string

// Error constants:
// ErrNotFound
// ErrConflict
```

The `key` parameter is the name of the blob. There are no key prefixes when operating with blobs: applications have full access to all data in the storage service. If `keyPrefix` is set in the metadata, it is ignored by `*Blob` methods.

The fact that there are no key prefixes is also one of the two reasons why `DeleteBlob` exists, as the `Delete` method is expected to respect key prefixes. The other reason is having consistent APIs, with similar arguments and usage.

Each blob can have metadata associated to it, which is normally passed to clients as headers (more on that below). In particular, `Content-Type` is yet another metadata key, treated no differently from anything else.

> Option: we could support auto-detection of the content type if users pass `"auto"` as value, using a library such as [ItalyPaleAle/file-type-stream-go](https://github.com/ItalyPaleAle/file-type-stream-go) (which I'd be very happy to transfer to the Dapr org).

The only exception is the `tag`, which is passed separately from the metadata objet. The reason is that in the `SetBlob` method, the tag is both an (optional) parameter and a return value. Note however that not all stores support tags, and it's not a requirement.

> When operating with blobs, Dapr only supports the "last-write-wins" concurrency pattern. This makes sense considered the intended usage of blob storage, which is storing documents that usually do not change. When documents do change, using last-write-wins is consistent with how regular filesystems work, for example.

When invoking `GetBlob` or `SetBlob`, the output and input streams (respectively) are created outside of the state store and are owned by the caller. The `GetBlob` and `SetBlob` methods synchronously return only after all data has been written or read (respectively), and after that the caller can call `Close()` on the streams (if necessary/appropriate).

Notes:

1. Automatic state encryption is not supported by blob storage methods either. The primary reason is that automatic state encryption is designed for individual applications to persist their own state, while blobs are meant to be used by a variety of applications and for various purposes. Applications that want to encrypt blobs can leverage the upcoming cryptography building block (see #4508) or implement custom solutions.
2. Dapr will not calculate and store the checksum of blobs. This is because checksums must be transmitted as header, but Dapr stores data in the state store in a streaming way, so it can't compute the full checksum until the end of the stream. Apps that want to store checksums (such as the `Content-MD5` header) should compute it beforehand and submit it as metadata header.

## HTTP APIs

Applications that interact with blob storage using HTTP can leverage these APIs, that follow a more REST-like pattern than the typical state store APIs.

> These APIs have been modeled after the APIs supported by AWS S3 and Azure Blob Storage.

### Retrieving a blob

To retrieve a blob, clients make a GET request:

```text
GET http://localhost:<port>/v1.0/blob/<state-store-name>/<key>
```

Where `<state-store-name>` is the name of the state store component, and `<key>` is the key (name) of the blob.

A successful response will be:

- Status: `200 OK`
- Headers: each metadata key is a header; see below for a comment on headers. Additionally, the `ETag` is passed as response header too.
- Body: the body of the response is the raw bytes of the blob, which are sent to the client in a streamed way, as soon as they are retrieved from the state store (that is: Dapr does not buffer the data in-memory before sending it to the client).

An error response will have:

- Status:
  - `404 Not Found` for blobs not found
  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
  - `500 Internal Server Error` for all other errors
- Headers: response will have `Content-Type: application/json`
- Body: a JSON-encoded payload including more details on the error

### Storing a blob

To store a blob, clients make a PUT request:

```text
PUT http://localhost:<port>/v1.0/blob/<state-store-name>/<key>
```

Where `<state-store-name>` is the name of the state store component, and `<key>` is the key (name) of the blob.

The request body contains the raw bytes of the blob, which are read by Dapr in a streamed way and are sent to the state store as soon as they are available (once again, Dapr does not buffer the data in-memory before storing it).

Additionally, metadata keys are passed as header values; see below for a comment on headers. An `ETag` can be passed as request header too for concurrency control, and it's ignored if the blob doesn't exist already. If an ETag is not specified, the write will always succeed (unless other errors occur).

A response will be:

- Status:
  - `200 OK` for successful operations
  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
  - `409 Conflict` if the blob already exists and there's an ETag mismatch
  - `500 Internal Server Error` for all other errors
- Headers: response will have `Content-Type: application/json`
- Body:
  - For successful responses: a JSON-encoded payload containing the ETag and the number of bytes stored.
  - In case of errors: a JSON-encoded payload including more details on the error

### Deleting a blob

To delete a blob, clients make a DELETE request:

```text
DELETE http://localhost:<port>/v1.0/blob/<state-store-name>/<key>
```

Where `<state-store-name>` is the name of the state store component, and `<key>` is the key (name) of the blob.

An `ETag` can be passed as request header too for concurrency control. If an ETag is is specified, the operation will fail if the ETag doesn't match; the ETag is optional and when missing there's no concurrency control.

A response will be:

- Status:
  - `204 No Content` for successful operations
  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
  - `409 Conflict` if the blob already exists and there's an ETag mismatch
  - `500 Internal Server Error` for all other errors
- Headers: error responses will have `Content-Type: application/json`
- Body:
  - For successful responses, the response body is empty
  - In case of errors: a JSON-encoded payload including more details on the error

## gRPC APIs

For applications using gRPC to interact with Dapr, the `Dapr` service is expanded to include:

```proto3
service Dapr {
  // GetBlob retrieves a blob.
  rpc GetBlob(GetBlobRequest) returns (stream GetBlobResponse) {}

  // SetBlob stores a blob with name key.
  rpc SetBlob(stream SetBlobRequest) returns (SetBlobResponse) {}

  // DeleteBlob a blob with name key.
  rpc DeleteBlob(DeleteBlobRequest) returns (DeleteBlobResponse) {}
}

// GetBlobRequest is the message to get a blob from specific state store.
message GetBlobRequest {
  // The name of state store.
  string store_name = 1;

  // The key of the desired blob.
  string key = 2;
}

// GetBlobResponse contains the retrieved blob (this is used in a streamed response).
message GetBlobResponse {
  // The tag for concurrency control.
  string tag = 1;

  // The metadata stored with the blob.
  map<string, string> metadata = 2;

  // Chunk of data.
  StreamPayload payload = 10;
}

// SetBlobRequest is the message to store a blob in a specific state store (this is used in a streamed request).
message SetBlobRequest {
  // The name of state store.
  string store_name = 1;

  // The key of the desired blob.
  string key = 2;

  // The tag for concurrency control.
  string tag = 3;

  // The metadata which will be stored with the blob.
  map<string, string> metadata = 4;

  // Chunk of data.
  StreamPayload payload = 10;
}

// SetBlobResponse contains the result of storing the blob.
message SetBlobResponse {
  // The updated tag for concurrency control.
  string tag = 1;

  // The number of bytes written.
  int bytes = 2;
}

// DeleteBlobRequest is the message to delete a blob in specific state store.
message DeleteBlobRequest {
  // The name of state store.
  string store_name = 1;

  // The key of the desired blob.
  string key = 2;
}

// DeleteBlobResponse contains the result of deleting the blob.
message DeleteBlobResponse {
    // Currently empty but allowing for future expansion.
}
```

### Handling streams

`StreamPayload` is first introduced with #4903 and corresponds to:

```proto3
// Chunk of data sent in a streaming request or response.
message StreamPayload {
  // Data sent in the chunk.
  google.protobuf.Any data = 1;

  // Set to true if this is the last chunk.
  bool complete = 2;
}
```

Data is sent from the application to Dapr (in `SetBlob` RPCs) or from Dapr to the application (in `GetBlob` RPCs) in a stream. Each message in the stream contains a `GetBlobResponse` or `SetBlobRequest` where:

- The first message in the stream MUST contain all other required keys.
- The first message in the stream MAY contain a `payload`, but that is not required.
- Subsequent messages (any message except the first in the stream) MUST contain a `payload` and MUST NOT contain any other property.
- The last message in the stream MUST contain a `payload` with `complete` set to `true`. That message is assumed to be the last one from the sender and no more messages are to be sent in the stream after that.

The amount of data contained in `payload.data` is variable and it's up to the discretion of the sender. In service invocation calls, as implemented by #4903, chunks are at most 4KB in size, although senders may send smaller chunks if they wish. Receivers must not assume that messages will contain any specific number of bytes in the payload.

Note that it's possible for senders to send a single message in the stream. If the data is small and could fit in a single chunk, senders MAY choose to include a `payload` with `complete=true` in the first message. Receivers should assume that single message to be the entire communication from the sender.

## Metadata and headers

Metadata properties are passed between the client and server as headers.

There can be two types of headers:

- Custom ones have the `x-dapr-` prefix in requests sent between Dapr and the app. When they are stored, they generally do not have the `x-dapr-` prefix, which is removed (although this could be dependent on the implementation of the state store).
- Certain common headers are exchanged between apps and Dapr in their canonical form. The way these are stored in the state store depends on the service and implementation:
  - `Last-Modified`
  - `Content-Length`
  - `Content-Type`
  - `Content-MD5`
  - `Content-Encoding`
  - `Content-Language`
  - `Cache-Control`
  - `Origin`
  - `Range`
  - Additionally, the `ETag` header is exchanged in the request/response headers, although it's not included in the metadata object.

**Comment by yaron2** [comment]
> GET http://localhost:<port>/v1.0/blob/<state-store-name>/<key>

Please change `v1.0/blob` to `v1.0/state/blog` (or any alternative convention that ties this to state) as the current design suggests this is a new API, which is what we agreed we wouldn't do based on the discussion in #4804.

**Comment by ItalyPaleAle** [comment]
@yaron2  I thought of that, but when I started writing the proposal, I realized there were 2 problems with it:

**First**, a conflict with the API to retrieve states, which is `GET http://localhost:<daprPort>/v1.0/state/<storename>/<key>`.

**Second** the blob API should behave differently from the state store API, including two things I listed above (for both technical and "philosophical" reasons):

1. The key is used as-is and the appID and/or namespace are not prepended to the key.
2. Things like automatic state store encryption are not leveraged.

IMHO it makes sense to have a different API here. I agreed on #4804 that we shouldn't create a new building block meaning that we shouldn't create a new class of components (like `blob.aws.s3` that would be in addition to `state.aws.s3` and `binding.aws.s3`... what a mess!). However, as I was designing the above, keeping the same `/state` prefix seemed it would have caused more confusion, because of ambiguity of paths and differences in behaviors.

**Comment by yaron2** [comment]
Ok, that *kind of* makes sense :)

I'm not sure about the API conflict as the router should resolve `/state/<dynamic>/<dynamic>` and `/state/blob/<dynamic>/<dynamic>` just fine.

I agree the main thing here is to not make this a new building block by qualifying a new component class and this proposal addresses that concern.

**Comment by yaron2** [comment]
> There are no key prefixes when operating with blobs

What's the reasoning behind that?

**Comment by ItalyPaleAle** [comment]
> I'm not sure about the API conflict as the router should resolve /state/<dynamic>/<dynamic> and /state/blob/<dynamic>/<dynamic> just fine.

I was concerned about the case where blob keys contain `/`, but looks like we don't support that scenario today anyways. So yes, this would not be a problem for GET (or POST) requests.

> >  There are no key prefixes when operating with blobs
>
> What's the reasoning behind that?

I understand that for state we add prefixes because we design that to be an app-specific thing: apps can store their own state and its meant for them only (there are ways to not have a prefix, but that's another topic). This is more aligned to how people use databases, most times: without Dapr, a developer would likely create a DB table for each application to store this kind of state.

In the case of blob storage, when I think about the use cases, in my mind they're primarily centered around storing large files that will be used by other applications. A canonical example would be storing photos (think a cat photo sharing website, or more realistically an e-commerce application, etc): once the file is in the blob storage, it may even be accessed directly by users through their browsers. This is more aligned with how people use S3, most times.

**Comment by yaron2** [comment]
> In the case of blob storage, when I think about the use cases, in my mind they're primarily centered around storing large files that will be used by other applications. A canonical example would be storing photos (think a cat photo sharing website, or more realistically an e-commerce application, etc): once the file is in the blob storage, it may even be accessed directly by users through their browsers. This is more aligned with how people use S3, most times.

Developers could simply set the `keyPrefix: none` setting on their component (example [here](https://docs.dapr.io/developing-applications/building-blocks/state-management/howto-share-state/#none)) to achieve the functionality you're describing and remain consistent with how Dapr handles state for non-blob entities. WDYT?

**Comment by ItalyPaleAle** [comment]
It's an option. I think it all depends on how users will want to use this, and on whether we use `v1.0/state` (in which case the closest we are to the current behavior, perhaps the better) or `v1.0/blob` for this (in which case we can make decisions more freely

Do you think this could use some UX research?

**Comment by yaron2** [comment]
> It's an option. I think it all depends on how users will want to use this, and on whether we use `v1.0/state` (in which case the closest we are to the current behavior, perhaps the better) or `v1.0/blob` for this (in which case we can make decisions more freely
> 
> Do you think this could use some UX research?

For sure.

**Comment by ItalyPaleAle** [comment]
Summoning our PM @nyemade-uversky :)

**Comment by olitomlinson** [comment]
I think it's misplaced to assume that developers won't want any isolation guarantees for blobs. I understand that a canonical use-case for large blobs is for sharing photos and possibly other _public_ content, however in my use-case (which is pharma/therapeutics) we often deal with large blobs of sensitive data, and our architecture would benefit greatly from layers of isolation that is out of the immediate control of the developer, such as what is possible with the `keyPrefix`

As a user, my two pence : keep to the same state sharing semantics (`none`, `name`, `namespace`,`appId`)  if possible. It also paints a consistent story with the existing state APIs, which makes it simpler for the user to reason about.

---

Not sure if I'm being dumb here, but I'm confused around concurrency control. One paragraph states :

> When operating with blobs, **Dapr only supports the "last-write-wins" concurrency pattern**. This makes sense considered the intended usage of blob storage, which is storing documents that usually do not change. When documents do change, using last-write-wins is consistent with how regular filesystems work, for example.

But later on in the proposal, `Etags` are mentioned for controlling concurrency, which I read as that LWW is not the _only_ concurrency pattern? Or have I got that wrong?

**Comment by ItalyPaleAle** [comment]
> But later on in the proposal, Etags are mentioned for controlling concurrency, which I read as that LWW is not the only concurrency pattern? Or have I got that wrong?

Perhaps I misspoke (mis-wrote? :) ) myself.

Yes, we will support concurrency control using ETags. But if the ETag isn't specified, then the write will always succeed. I may have misunderstood how FWW (and that's totally possible), but I thought that with FWW, no etag would make the write fail.

**Comment by ItalyPaleAle** [comment]
Thanks also for sharing your input on prefixes. I think that there are enough reasons to implement support for prefixes (and it's a relatively low-cost thing to do).

What I think we could still debate on is what the default value should be. Should we behave like "regular" state stores and keep a prefix on by default, or should that be opt-in?

**Comment by ItalyPaleAle** [comment]
Thanks also for sharing your input on prefixes. I think that there are enough reasons to implement support for prefixes (and it's a relatively low-cost thing to do).

What I think we could still debate on is what the default value should be. Should we behave like "regular" state stores and keep a prefix on by default, or should that be opt-in?

Perhaps @nyemade-uversky or @paulyuk have thoughts on this

**Comment by yaron2** [comment]
> What I think we could still debate on is what the default value should be. Should we behave like "regular" state stores and keep a prefix on by default, or should that be opt-in?

Consistency is key for user experience. I highly encourage to keep the same defaults and allow users to override based on their individual need.

**Comment by berndverst** [comment]
Can we add `List` as an operation? This appears to be something that folks want to do for the majority of blob stores. (And of course the list operation could be keyPrefix aware)

@yaron2 @ItalyPaleAle @olitomlinson

**Comment by ItalyPaleAle** [comment]
Yes, that makes sense. With `list`, however, we'd need to implement some pagination logic.

**Comment by berndverst** [comment]
> Yes, that makes sense. With `list`, however, we'd need to implement some pagination logic.

Agreed. In many cases we can rely on the SDK's existing pagers, but in other cases we need to have a Dapr implementation for when native pagination does not exist.

**Comment by artursouza** [comment]
I am in favor of having more specialized building blocks vs bundling everything into state store type of design. We are doing that for Document Store and this fits in the same type of variation. I vote for not to bundle more building blocks into state store and add more to our feature compatibility matrix problem.

I still prefer this approach: https://github.com/dapr/dapr/issues/4804

**Comment by ItalyPaleAle** [comment]
Closing in favor of dapr/proposals#18 (which goes back to a new building block)
